experimental/storage/src/upload/types.ts

```diff
@@ -0,0 +1,68 @@
+import { PutObjectCommandInput, S3, S3Client, Tag as Tag_ } from "@aws-sdk/client-s3";
+
+export namespace Upload {
+  export interface Progress {
+    loaded?: number;
+    total?: number;
+    part?: number;
+    Key?: string;
+    Bucket?: string;
+  }
+
+  export interface S3Location {
+    Key: string;
+    Bucket: string;
+  }
+
+  export interface S3Target extends S3Location {
+    service: S3;
+  }
+  export interface S3ClientTarget extends S3Location {
+    service: S3Client;
+  }
+
+  export type Target = S3Target | S3ClientTarget;
```

Comment:
Why not consolidate to:

```typescript
export type Target = {
  service: S3 | S3Client;
  Key: string;
  Bucket: string;
}
```

Reply:
Yeah we totally could. But thats saying all targets have Keys and Buckets. Is that a door we want to enter?